### ISSUE TITLE
502 error on deposit

### ISSUE THREAD

**Issue by JosephMcArthur** [oaworks/discussion#1852]
With `10.1021/acs.jmedchem.1c00837` we're seeing:

`Sorry, we were not able to deposit this paper for you. Please try contacting us directly`

on attempts to deposit on live & dev.

When looking at dev tools I see:

![image](https://user-images.githubusercontent.com/4953703/158211332-2ae8ca83-74ef-4e39-8b1a-a888e6581e36.png)

I also see the following URL appear in the network tab:

https://beta.oa.works/ping?action=shareyourpaper_couldnt_submit_deposit&from=4WDY6njiatg3FZMJm&url=https%3A%2F%2Fdev.shareyourpaper.org%2F10.1021%2Facs.jmedchem.1c00837&live=1585895131684

The error seems to have first occurred yesterday (is that right @swipperman). Then @swipperman was able to deposit normally. Now, when I try on dev I see the same error as the original user.

User & sarah tried to deposit via: https://ShareYourPaper.org/funders/gates/10.1021/acs.jmedchem.1c00837?ref=email&email=kelly.chibale@uct.ac.za

The file is here: [1,3-Diarylpyrazolyl-acylsulfonamides as potent anti-tuberculosis agents.docx](https://github.com/oaworks/discussion/files/8246404/1.3-Diarylpyrazolyl-acylsulfonamides.as.potent.anti-tuberculosis.agents.docx)

**Comment by swipperman** [comment]
Yes, that is correct. The user contacted me on Saturday with a deposit question that I answered, then again on Sunday with the error. I'm assuming the user tried to deposit on Sunday and wrote me right after - the message was marked Mar 13, 2022, 12:57 PM EST.

**Comment by markmacgillivray** [comment]
Trying to pass this file myself to the software we use to extract the text from docx files causes it to process indefinitely. So the API app times out and returns the user an error. I think this is probably the best solution for this case, as then they come to us and ask for manual help, which is all we can do here. docx are always a bit fiddly.

**Comment by JosephMcArthur** [comment]
Gotcha, so you think this is a one off (or at least, extremely infrequent)
error caused by file weirdness?

**Comment by markmacgillivray** [comment]
I don't know how infrequent it will be, we do know that the tools we have worked for the docx files we tested for, but we'll just have to see how many docx files cause a problem over time. If it's too many we'll need to look for an alternative approach to getting text out of them. But yes it's caused by file weirdness.

**Comment by swipperman** [comment]
That's really helpful info for support on my end, too. For troubleshooting, would it then be best to try a PDF if the docx file isn't working?

**Comment by JosephMcArthur** [comment]
Ah, interesting, yes converting to a PDF does actually resolve this issue! Sarah, did you do that in your successful attempt to deposit it?

Mark, in cases like this is there any way for us to send a review deposit email instead of just failing?

**Comment by swipperman** [comment]
I did not convert it in that case - the docx just worked for me!

**Comment by markmacgillivray** [comment]
We could add a timeout, but then we risk timing out ones that actually would succeed

**Comment by JosephMcArthur** [comment]
Can you help me understand how a timeout helps please?

**Comment by markmacgillivray** [comment]
When we try to extract text from a document, we hand it off to a text extraction process. We hope it returns within a reasonable time. Then then deposit process continues, analysing the text, then returning a result. However, if the text extraction takes so long that the deposit process takes too long as well, then the web server (nginx) will timeout the connection. Then the embed handles that and tells the user to contact us.

If we add a timeout on the deposit process itself, say two minutes, the timer would start running when the deposit process begins. If it is still running once the timeout is reached, it could then stop in a controlled manner and do whatever you want it to do - for example, create a review deposit.

The downside is that it then WILL timeout at whatever timeout value we set, even for ones that may otherwise have succeeded just beyond that timeout. It's hard to balance timeouts because they matter for service responsiveness and also depend on what sorts of files we are given, and their content.

**Comment by markmacgillivray** [comment]
Is _oab_check what would normally happen for a review deposit? If so then this is just an API-side change, and would take a couple of hours. If there needs to be different behaviour on the embed, then it's maybe a couple of hours more, and probably relates to other embed changes in relation to review statuses.

**Comment by JosephMcArthur** [comment]
> Is _oab_check what would normally happen for a review deposit?

Partly. Normal they go via `_oab_confirm`, but I don't know if that matters here –– I think you'd need to check.

And, yes, I think it probably does relate to https://github.com/oaworks/discussion/issues/1847 a bit, but sadly also https://github.com/oaworks/discussion/issues/1846(as otherwise the test will fail shortly). Everything is always connected to everything 😭.

Sadly I fear we should hold off until we've done more of https://github.com/oaworks/internal-planning/issues/265 is done.

**Comment by swipperman** [comment]
In case this is of use, I had another author who got the timeout with a docx. I tried it myself and had the same issue. I converted to a PDF, and it was sent for review, which I then deposited.

**Comment by JosephMcArthur** [comment]
Definitely of use! Can you share the file and doi?

**Comment by swipperman** [comment]
Here's where it was deposited: https://zenodo.org/record/6366709

docx and pdf attached

[2. Reviewed_Diversity of culturable bacteria isolated from the feces of wild Anopheles darlingi (Arruda et al) 11.01.2021_final.pdf](https://github.com/oaworks/discussion/files/8302653/2.Reviewed_Diversity.of.culturable.bacteria.isolated.from.the.feces.of.wild.Anopheles.darlingi.Arruda.et.al.11.01.2021_final.pdf)
[2. Reviewed_Diversity of culturable bacteria isolated from the feces of wild Anopheles darlingi (Arruda et al) 11.01.2021_final.docx](https://github.com/oaworks/discussion/files/8302655/2.Reviewed_Diversity.of.culturable.bacteria.isolated.from.the.feces.of.wild.Anopheles.darlingi.Arruda.et.al.11.01.2021_final.docx)

**Comment by markmacgillivray** [comment]
Is there anything else I should be doing here?

**Comment by JosephMcArthur** [comment]
Not right now, no. Unassigned.

**Comment by JosephMcArthur** [comment]
Another timeout reported for:

* 10.1111/jora.12689
* [crosnoe_final.docx](https://github.com/oaworks/discussion/files/8326628/crosnoe_final.docx)

That's three now, out of about 20ish gates deposits in the past few weeks.

That makes this a fairly significant issue worthy of attention.

Mark, can you advise on two things:

> Partly. Normal they go via _oab_confirm, but I don't know if that matters here –– I think you'd need to check.

1. Can you please check if a frontend change (as above) would be required to cleanly handle the timeout?
2. What are our options to fix the "root" cause here i.e whatever is causing the timeout. Do we know, and just haven't done it, or do we need to investigate?

Note: this isn't more important than existing assigned Gates work. Let's polish up those issues and then move back to this.

**Comment by markmacgillivray** [comment]
Our new API uses different underlying software to do the docx extraction because the stuff the old one used was no longer supported and a bit of a problem to install on newer systems. However if we're seeing too-often issues with the new underlying software we can put some time into trying to debug it and/or find a different alternative. 

We have an ability to send an error response from the deposit endpoint to the embed already. In this case the deposit type is set to review and the review email gets sent. The embed then shows a specific panel for that, called _oab_partial. The content of it is: "Congrats, you\'re done! Check back soon to see your paper live, or we'll email you with issues." If that is acceptable, then I can implement the 60 second timeout on the backend, send the error response if it is triggered, and the embed will show that partial done state.

**Comment by JosephMcArthur** [comment]
> We have an ability to send an error response from the deposit endpoint to the embed already. In this case the deposit type is set to review and the review email gets sent. The embed then shows a specific panel for that, called _oab_partial. The content of it is: "Congrats, you're done! Check back soon to see your paper live, or we'll email you with issues." If that is acceptable, then I can implement the 60 second timeout on the backend, send the error response if it is triggered, and the embed will show that partial done state.

Thanks for looking into this. I'd like us to implement this as you suggest first and soon assuming we still think it's a couple of hours more work. Then, with that fix live, let's explore how to address the root cause.

**Comment by JosephMcArthur** [comment]
This should've been here: https://github.com/oaworks/discussion/issues/1852

**Comment by JosephMcArthur** [comment]
If we could push this live today that would be fab

**Comment by JosephMcArthur** [comment]
Tested quickly on our call today. The new file above timed out, sent an email, and displayed on the UI as expected.

Can test using:
* https://beta.oa.works/archivable?url=https://github.com/oaworks/api/files/8364525/DTG.paper_rev_JD2_TO2_unmarked.docx
* https://beta.oa.works/archivable?timeout=1

which shows the timeout state as expected.

So, I need to:

- [ ] Make an issue for future follow up on root cause
- [ ] Add a test

**Comment by markmacgillivray** [comment]
Live has been updated

**Comment by JosephMcArthur** [comment]
Thanks Mark!

Joseph McArthur
Assistant Director: Right to Research Coalition <http://righttoresearch.org>
Director / Founder: Open Access Button <http://openaccessbutton.org>
Twitter: @Mcarthur_Joe <https://twitter.com/Mcarthur_Joe>
Skype: joseph_mcarthur



On Wed, Apr 06, 2022 at 10:05 AM, markmacgillivray ***@***.***

**Comment by swipperman** [comment]
Update: We have had some more timeouts reported for Gates authors without receiving deposit check emails. 

One author said she tried uploading a few times on Apr 7 to https://shareyourpaper.org/funders/gates/10.1080/09614524.2021.2009772?ref=email&email=jamelia.harris@fiscus.org.uk without success. We did not receive any deposit check emails, so I'm assuming it timed out. The author's help request went into the spam folder, so I just tried it yesterday - iirc, it timed out for with the .docx file but worked as a .pdf. 

Another author wrote today:  "I have tried to deposit the paper in the past (and again today) but keep receiving the error “Sorry, we were not able to deposit this paper for you. Please try contacting us directly”." However, the paper went through this morning, and we were sent a deposit check email. Again, it timed out for me as a .docx but worked as a .pdf. That one was for https://shareyourpaper.org/funders/gates/10.1002/hec.4367?ref=email&email=r.a.thomas@lse.ac.uk

**Comment by JosephMcArthur** [comment]
Thanks! Can you share the files @swipperman?

**Comment by swipperman** [comment]
Here are the files: 
[HEC-20-0382.R2 Manuscript_synthetic_control_Health Economics_final_Gates.docx](https://github.com/oaworks/discussion/files/8522417/HEC-20-0382.R2.Manuscript_synthetic_control_Health.Economics_final_Gates.docx)
[Specific Purpose Grants_manuscriptR1.docx](https://github.com/oaworks/discussion/files/8522420/Specific.Purpose.Grants_manuscriptR1.docx)

**Comment by markmacgillivray** [comment]
Possibly related to #1842

**Comment by JosephMcArthur** [comment]
Seeing more reports of this, so while I think you've already got this near the top of your list I just want to explicitly say please have this higher than work on OAreport / other fixes

**Comment by markmacgillivray** [comment]
As suspected in my earlier comment, this appears to be related to the reversion of the embed code. So I've been continuing with #1842 on the assumption it will solve this for us as well (solve in the sense of getting us back to having a more useful timeout that tells us an attempt was made).

**Comment by markmacgillivray** [comment]
I've altered dev to serve the new embed, and confirmed it is back to working with the deposit, and uploading one of the above files results in the expected new error response from the backend, and shows the expected conclusion to the user. I've also written the same alterations to live, and can turn those on any time - do you want to take a look at dev to confirm this, or will I go ahead?

This time I've handled the serving of the new embed via the nginx config, so it can't be broken by any future website updates. We already have an issue open to track moving the old website to a new server (whenever we move the new API to an oa.works DO account) or to a netlify account, so these nginx configs will do until then, and will also be usable if we do just stick the old website on the new API server as well.

**Comment by JosephMcArthur** [comment]
Let's go ahead and push it live if you're confident!

Joseph McArthur
Assistant Director: Right to Research Coalition <http://righttoresearch.org>
Director / Founder: Open Access Button <http://openaccessbutton.org>
Twitter: @Mcarthur_Joe <https://twitter.com/Mcarthur_Joe>
Skype: joseph_mcarthur

**Comment by markmacgillivray** [comment]
I have found an issue with pushing it to live to do with the config of instantill.org and shareyourpaper.org domains, which are different from how they are in dev. This does not break anything and isn't a problem with the intended solution, it just means it'll take me a bit longer to finish this because I have to find which other old domain routing is overwriting the new change we need for the new embed. I'll close this issue once I've found it and finished this.
